Mathlib/LinearAlgebra/Span/Basic.lean

```diff
@@ -162,26 +162,72 @@ theorem disjoint_span_singleton' {K E : Type*} [DivisionRing K] [AddCommGroup E]
     {p : Submodule K E} {x : E} (x0 : x ≠ 0) : Disjoint p (K ∙ x) ↔ x ∉ p :=
   disjoint_span_singleton.trans ⟨fun h₁ h₂ => x0 (h₁ h₂), fun h₁ h₂ => (h₁ h₂).elim⟩
 
-variable (R S s)
+section IsScalarTower
+
+variable (S)
+
+variable [Semiring S] [SMul R S] [Module S M] [IsScalarTower R S M] (p : Submodule R M)
+
+/-- The inclusion of an `R`-submodule into its `S`-span, as an `R`-linear map. -/
+@[simps] def inclusionSpan :
+    p →ₗ[R] span S (p : Set M) where
+  toFun x := ⟨x, subset_span x.property⟩
+  map_add' x y := by simp
+  map_smul' t x := by simp
+
+lemma injective_inclusionSpan :
+    Injective (p.inclusionSpan S) := by
+  intro x y hxy
+  rw [Subtype.ext_iff] at hxy
+  simpa using hxy
+
+lemma span_range_inclusionSpan :
+    span S (range <| p.inclusionSpan S) = ⊤ := by
+  have : (span S (p : Set M)).subtype '' range (inclusionSpan S p) = p := by
+    ext; simpa [Subtype.ext_iff] using fun h ↦ subset_span h
+  apply map_injective_of_injective (span S (p : Set M)).injective_subtype
+  rw [map_subtype_top, map_span, this]
+
+variable (R s)
 
 /-- If `R` is "smaller" ring than `S` then the span by `R` is smaller than the span by `S`. -/
-theorem span_le_restrictScalars [Semiring S] [SMul R S] [Module S M] [IsScalarTower R S M] :
+theorem span_le_restrictScalars :
     span R s ≤ (span S s).restrictScalars R :=
   Submodule.span_le.2 Submodule.subset_span
 
 /-- A version of `Submodule.span_le_restrictScalars` with coercions. -/
 @[simp]
-theorem span_subset_span [Semiring S] [SMul R S] [Module S M] [IsScalarTower R S M] :
+theorem span_subset_span :
     ↑(span R s) ⊆ (span S s : Set M) :=
   span_le_restrictScalars R S s
 
 /-- Taking the span by a large ring of the span by the small ring is the same as taking the span
 by just the large ring. -/
-theorem span_span_of_tower [Semiring S] [SMul R S] [Module S M] [IsScalarTower R S M] :
+@[simp]
+theorem span_span_of_tower :
     span S (span R s : Set M) = span S s :=
   le_antisymm (span_le.2 <| span_subset_span R S s) (span_mono subset_span)
 
-variable {R S s}
+variable {R S} in
+lemma span_range_inclusion_eq_top (p : Submodule R M) (q : Submodule S M)
+    (h₁ : p ≤ q.restrictScalars R) (h₂ : q ≤ span S p) :
+    span S (range (inclusion h₁)) = ⊤ := by
+  suffices (span S (range (inclusion h₁))).map q.subtype = q by
+    apply map_injective_of_injective q.injective_subtype
+    rw [this, q.map_subtype_top]
+  rw [map_span]
+  suffices q.subtype '' ((LinearMap.range (inclusion h₁)) : Set <| q.restrictScalars R) = p by
+    refine this ▸ le_antisymm ?_ h₂
+    simpa using span_mono (R := S) h₁
+  ext x
+  simpa [range_inclusion] using fun hx ↦ h₁ hx
+
+@[simp]
+theorem span_range_inclusion_restrictScalars_eq_top :
+    span S (range (inclusion <| span_le_restrictScalars R S s)) = ⊤ :=
+  span_range_inclusion_eq_top _ _ _ <| by simp
```

Comment:
These lemmas are remnants from an earlier (messier) implementation. They seem just about worth keeping but I'd be happy to drop them if others aren't convinced.